wasm/validator/src/visitor.rs

```diff
@@ -0,0 +1,747 @@
+use alloc::borrow::ToOwned as _;
+
+use iroha_wasm::data_model::{transaction::TransactionPayload, NumericValue, SignaturesOf};
+
+use super::prelude::*;
+use crate::data_model::prelude::*;
+
+macro_rules! delegate {
+    ( $($validator:ident($operation:ty)),+ $(,)? ) => { $(
+        #[allow(non_snake_case)]
+        fn $validator(&mut self, authority: &AccountId, operation: $operation) -> Verdict {
+            $validator(self, authority, operation)
+        } )+
+    }
+}
+
+macro_rules! deny_unsupported_instruction {
+    ($isi_type:ty) => {
+        deny!(concat!("Unsupported `", stringify!($isi_type), "` instruction").to_owned())
+    };
+}
+
+macro_rules! evaluate_field {
+    ($validator:ident, $authority:ident, <$isi:ident as $isi_type:ty>::$field:ident) => {{
+        fn type_check(_isi: &$isi_type) {}
+        type_check($isi);
+
+        let verdict = validate_expression($validator, $authority, $isi.$field().expression());
+        if verdict.is_deny() {
+            return verdict;
+        }
+
+        $isi.$field().evaluate(&Context::new()).dbg_expect(concat!(
+            "Failed to evaluate `",
+            stringify!($field),
+            "` of `",
+            stringify!($isi_type),
+            "`"
+        ))
+    }};
+}
+
+/// Trait to validate Iroha entities.
+///
+/// This trait is based on the visitor pattern
+pub trait Validate {
+    delegate! {
+        // Validate SignedTransaction
+        validate_transaction(SignedTransaction),
+        validate_transaction_signatures(&SignaturesOf<TransactionPayload>),
+        validate_transaction_payload(&TransactionPayload),
+
+        // Validate TransactionPayload
+        validate_instruction(&InstructionBox),
+        validate_wasm(&WasmSmartContract),
+        validate_expression(&Expression),
+        validate_query(&QueryBox),
+
+        // Validate InstructionBox
+        validate_burn(&BurnBox),
+        validate_execute_trigger(&ExecuteTrigger),
+        validate_fail(&FailBox),
+        validate_grant(&GrantBox),
+        validate_if(&Conditional),
+        validate_mint(&MintBox),
+        validate_new_parameter(&NewParameter<Parameter>),
+        validate_pair(&Pair),
+        validate_register(&RegisterBox),
+        validate_remove_key_value(&RemoveKeyValueBox),
+        validate_revoke(&RevokeBox),
+        validate_sequence(&SequenceBox),
+        validate_set_key_value(&SetKeyValueBox),
+        validate_set_parameter(&SetParameter<Parameter>),
+        validate_transfer(&TransferBox),
+        validate_unregister(&UnregisterBox),
+        validate_upgrade(&UpgradeBox),
+
+        // Validate QueryBox
+        validate_DoesAccountHavePermissionToken(&DoesAccountHavePermissionToken),
```

Comment:
Why did you choose to use `CamelCase` for queries while using `snake_case` for all other places?

Reply:
it's only for the queries because:
* it's more readable because queries are always made up of multiple words
* you can immediately identify it's a query you are validating
* I was lazy

consider for queries:
* `validate_does_account_have_permission_token_query`
* `validate_query_does_account_have_permission_token`
* `validate_does_account_have_permission_token`
* `validate_DoesAccountHavePermissionToken`

consider for ISI:
* `validate_register_isi`
* `validate_isi_register`
* `validate_register`
* `validate_Register`

can we agree on some convention out of these? or do you have a different suggestion?

Reply:
I'm okay with `validate_does_account_have_permission_token` variant

Reply:
ok, will make it snake case